src/out_of_order_engine.cc

```diff
@@ -0,0 +1,414 @@
+#include "out_of_order_engine.h"
+#include "dense_map.h"
+#include "instruction_graph.h"
+#include "system_info.h"
+#include "utils.h"
+
+#include <queue>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+#include <matchbox.hh>
+
+namespace celerity::detail::out_of_order_engine_detail {
+
+using target = out_of_order_engine::target;
+using lane_id = out_of_order_engine::lane_id;
+using assignment = out_of_order_engine::assignment;
+
+/// Comparision operator to make priority_queue<instruction*> return concurrent instructions in decreasing priority.
+struct instruction_priority_less {
+	bool operator()(const instruction* lhs, const instruction* rhs) const { return lhs->get_priority() < rhs->get_priority(); }
+};
+
+/// Instruction is not eligible for assignment yet. Implies `num_incomplete_predecessors > 0`.
+struct unassigned_state {
```

Comment:
This type name uses `assigned`, while the others use `assignable`. The comment also talks in terms of being *eligible* for assignment. Should this be `unassignable_state`, or is there an actual semantic implication?

Reply:
After reading more of the implementation, I think the type name is correct but maybe the comment should be updated. As far as I can tell, an instruction in this state isn't necessarily "not eligible for assignment yet", it just hasn't been assigned yet?

Reply:
There is a difference:

- `unassigned` means "not ready for immediate or eager assignment"
- `eagerly assignable` means "can be submitted to an in-order queue / thread queue to implicitly fulfill dependencies, but it is conditional (now renamed for that reason)
- `assigned` means "has been returned from `assign_one`" and is waiting for a call to `complete_assigned`.

I improved the documentation.